app/services/knowledge_base/chunking.py

```
@@ -0,0 +1,248 @@
+            else:
+                paragraphs.append(para)
+
+        buf: List[str] = []
+        buf_tokens = 0
+
+        def flush_buf():
+            nonlocal buf, buf_tokens
+            if not buf:
+                return
+            body_text = "\n\n".join(buf)
+            prefix = f"Section: {crumb}\n\n" if crumb else ""
```

Ruff (0.15.20)

[warning] Function definition does not bind loop variable `crumb`

(B023)